runner/src/recipe-manager.ts

```diff
@@ -0,0 +1,327 @@
+ * 2. Preserves recipe IDs: Maintains consistency between local and remote IDs
+ * 3. Clear publishing flow: Only syncs with Blobby when explicitly requested
+ * 4. Attempts to download recipes from Blobby if no cell is found
+ * 5. Minimize requirements that Blobby is available for a space to run recipes
+ *
+ * Storage layers:
+ * - In-memory cache: Fast access during runtime
+ * - Cell storage: Persistent local storage
+ * - Blobby storage: Remote storage for sharing recipes
+ *
+ * Usage:
+ * - Use the singleton instance exported as `recipeManager`
+ * - For new code, prefer the `recipeManager` object
+ */
+
+import { JSONSchema, Module, Recipe, Schema } from "@commontools/builder";
+import { storage } from "./storage.ts";
+import { Cell } from "./cell.ts";
+import { createRef } from "./doc-map.ts";
+import { getCell } from "./cell.ts";
+import { runtime } from "@commontools/runner";
+import { getBlobbyServerUrl } from "./blobby-storage.ts";
+
+const inProgressCompilations = new Map<string, Promise<Recipe>>();
+
+// Schema definitions
+export const recipeMetaSchema = {
+  type: "object",
+  properties: {
+    id: { type: "string" },
+    src: { type: "string" },
+    spec: { type: "string" },
+    parents: { type: "array", items: { type: "string" } },
+    recipeName: { type: "string" },
+  },
+  required: ["id"],
+} as const satisfies JSONSchema;
+
+export type RecipeMeta = Schema<typeof recipeMetaSchema>;
+
+// Type guard to check if an object is a Recipe
+function isRecipe(obj: Recipe | Module): obj is Recipe {
+  return "result" in obj && "nodes" in obj;
+}
+
+// FIXME(ja): what happens when we have multiple active spaces... how do we make
+// sure we register the same recipeMeta in multiple spaces?
+const recipeMetaMap = new WeakMap<Recipe, Cell<RecipeMeta>>();
+const recipeIdMap = new Map<string, Recipe>();
+
+class RecipeManager {
+  private async getRecipeMetaCell(
+    { recipeId, space }: { recipeId: string; space: string },
+  ) {
+    const cell = getCell(
+      space,
+      { recipeId, type: "recipe" },
+      recipeMetaSchema,
+    );
+
+    await storage.syncCell(cell);
+    await storage.synced();
+    return cell;
+  }
+
+  // returns the recipeMeta for a loaded recipe
+  getRecipeMeta(
+    input: Recipe | Module | { recipeId: string },
+  ): RecipeMeta {
+    if ("recipeId" in input) {
+      const recipe = this.recipeById(input.recipeId);
+      if (!recipe) throw new Error(`Recipe ${input.recipeId} not loaded`);
+      return recipeMetaMap.get(recipe)?.get()!;
+    }
+    return recipeMetaMap.get(input as Recipe)?.get()!;
+  }
+
+  generateRecipeId(recipe: Recipe | Module, src?: string) {
+    let id = recipeMetaMap.get(recipe as Recipe)?.get()?.id;
+    if (id) {
+      console.log("generateRecipeId: existing recipe id", id);
+      return id;
+    }
+
+    id = src
+      ? createRef({ src }, "recipe source").toString()
+      : createRef(recipe, "recipe").toString();
```

Comment:
Is this how we have always done this?

Reply:
yep, I had only referenced it as "hmm" because iirc, createRef of the recipe graph can be expensive... @seefeldb said something like that (perhaps it wasn't recipes?)